File: js/src/maplestory/api/dto/character/characterBasicDto.ts
import { CharacterBasicDtoBody } from '@src/maplestory/api/response/character/characterBasicDtoBody';

/**
 * 캐릭터 기본 정보
 */
class CharacterBasicDto {
  date: Date | null;

  /**
   * 캐릭터 명
   */
  characterName: string;

  /**
   * 월드 명
   */
  worldName: string;

  /**
   * 캐릭터 성별
   */
  characterGender: string;

  /**
   * 캐릭터 직업
   */
  characterClass: string;

  /**
   * 캐릭터 전직 차수
   */
  characterClassLevel: string;

  /**
   * 캐릭터 레벨
   */
  characterLevel: number;

  /**
   * 현재 레벨에서 보유한 경험치
   */
  characterExp: number;

  /**
   * 현재 레벨에서 경험치 퍼센트
   */
  characterExpRate: string;

  /**
   * 캐릭터 소속 길드 명
   */
  characterGuildName: string | null;

  /**
   * 캐릭터 외형 이미지
   */
  characterImage: string;

  /**
   * 캐릭터 생성일
   */
  characterDateCreate: Date;

  /**
   * 최근 7일간 접속 여부
   */
  accessFlag: 'true' | 'false';

  /**
   * 해방 퀘스트 완료 여부
   */
  liberationQuestClearFlag: 'true' | 'false';

  constructor(obj: CharacterBasicDtoBody) {
    const {
      date,
      character_name,
      world_name,
      character_gender,
      character_class,
      character_class_level,
      character_level,
      character_exp,
      character_exp_rate,
      character_guild_name,
      character_image,
      character_date_create,
      access_flag,
      liberation_quest_clear_flag,
    } = obj;

    this.date = date ? new Date(date) : null;
    this.characterName = character_name;
    this.worldName = world_name;
    this.characterGender = character_gender;
    this.characterClass = character_class;
    this.characterClassLevel = character_class_level;
    this.characterLevel = character_level;
    this.characterExp = character_exp;
    this.characterExpRate = character_exp_rate;
    this.characterGuildName = character_guild_name;
    this.characterImage = character_image;
    this.characterDateCreate = new Date(character_date_create);
    this.accessFlag = access_flag;
    this.liberationQuestClearFlag = liberation_quest_clear_flag;
  }

  /**
   * 최근 7일간 접속 여부
   */
  get isAccessFlag() {
    return this.accessFlag === 'true';
  }

  /**
   * 해방 퀘스트 완료 여부
   */
  get isLiberationQuestClearFlag() {
    return this.liberationQuestClearFlag === 'true';
  }
}

export { CharacterBasicDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/msea/dto/character/CharacterBasicDTO.java
package dev.spiralmoon.maplestory.api.msea.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import dev.spiralmoon.maplestory.api.Utils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.ZonedDateTime;

/**
 * Character basic information
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterBasicDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterBasicDTO {

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    @JsonProperty("date")
    private String date;

    /**
     * Character name
     */
    @JsonProperty("character_name")
    private String characterName;

    /**
     * World name
     */
    @JsonProperty("world_name")
    private String worldName;

    /**
     * Character gender
     */
    @JsonProperty("character_gender")
    private String characterGender;

    /**
     * Character job
     */
    @JsonProperty("character_class")
    private String characterClass;

    /**
     * Character job advancement tier
     */
    @JsonProperty("character_class_level")
    private String characterClassLevel;

    /**
     * Character level
     */
    @JsonProperty("character_level")
    private long characterLevel;

    /**
     * Experience points gained at the current level
     */
    @JsonProperty("character_exp")
    private long characterExp;

    /**
     * Percentage of experience points at the current level
     */
    @JsonProperty("character_exp_rate")
    private String characterExpRate;

    /**
     * Guild name the character belongs to
     */
    @JsonProperty("character_guild_name")
    private String characterGuildName;

    /**
     * Character appearance image
     */
    @JsonProperty("character_image")
    private String characterImage;

    public void setCharacterImage(String characterImage) {
        this.characterImage = Utils.removeQuery(characterImage);
    }

    /**
     * Character creation date (SGT, daily data with hours and minutes set to 0)
     */
    @JsonProperty("character_date_create")
    private String characterDateCreate;

    /**
     * Login status in the past 7 days (true:logged in, false:not logged in)
     */
    @JsonProperty("access_flag")
    private String accessFlag;

    /**
     * Liberation quest completion status (true:completed, false:not completed)
     */
    @JsonProperty("liberation_quest_clear_flag")
    private String liberationQuestClearFlag;

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    public ZonedDateTime getDate() {
        return date != null
                ? Utils.toZonedDateTime(this.date)
                : null;
    }

    /**
     * Character creation date (SGT, daily data with hours and minutes set to 0)
     */
    public ZonedDateTime getCharacterDateCreate() {
        return characterDateCreate != null
                ? Utils.toZonedDateTime(this.characterDateCreate)
                : null;
    }

    /**
     * Login status in the past 7 days
     */
    public boolean isAccessFlag() {
        return "true".equals(this.accessFlag);
    }

    /**
     * Liberation quest completion status
     */
    public boolean isLiberationQuestClearFlag() {
        return "true".equals(this.liberationQuestClearFlag);
    }
}
